### ISSUE TITLE
Proposal: Moving Shipping Zones into core

### ISSUE THREAD

**Issue by mikejolley** [woothemes/woocommerce#8294]
## Preface

Before going into depth, let us look at the existing shipping method "Flat Rate Shipping".

![2015-06-05 at 10 25](https://cloud.githubusercontent.com/assets/90977/8003202/4af32628-0b6d-11e5-873d-6705fe6c19dd.png)

This is an absolute frankenstein of a once simple method, now bloated with "features". Fat Rate Shipping. We'll get back to this later.

Like most of the shipping methods, it has a country selection option which allows you to **link a shipping method to a set of countries**.

![2015-06-05 at 10 27](https://cloud.githubusercontent.com/assets/90977/8003232/98c5f36c-0b6d-11e5-8479-6faa2c9694e0.png)

International shipping (a renamed flat rate) has the same thing; same options + a country selector.

My proposal is largely about improving the way you assign shipping methods to countries, but also a way to simplify the above methods.

## Enter Zones

Shipping Zones is something I built a long time ago for Table Rate Shipping. It lets you define groups of locations in which you offer shipping.

TL;DR A shipping zone is a group of either countries, states, or zip codes.

Here is the current zones UI: 

![2015-06-05 at 10 31](https://cloud.githubusercontent.com/assets/90977/8003301/295ea874-0b6e-11e5-8a32-342c0940d207.png)

Once a zone is added, table rate shipping lets you add multiple instances of itself to each zone:

![2015-06-05 at 10 38](https://cloud.githubusercontent.com/assets/90977/8003386/098c6b16-0b6f-11e5-8e66-6592715ebab4.png)

Now consider how this could benefit other methods..

## Flat Rate Reborn

If we were to merge zones into core Flat Rate Shipping would be so much more useful. Take out the bloat - just offer a single cost field and boom you could add multiple flat rates to multiple zones to make it much more flexible. "international shipping" could be dropped.

Example use case:

```
Zone 1 = UK and Ireland
Zone 2 = All of Europe
Zone 3 = US

Zone 1 Rates:
- Flat Rate named 'standard delivery' costing £5
- Flat Rate named 'rushed delivery' costing £8

Zone 2 Rates:
- Flat Rate named 'EU delivery' costing £15

Zone 3 Rates:
- The USPS extension, just for US
```

## Settings

Current settings all fall under the 'shipping' tab in admin:

![2015-06-05 at 10 51 2x](https://cloud.githubusercontent.com/assets/90977/8003589/d28c7adc-0b70-11e5-81c4-aa220ab38fa4.png)

Top level settings should be supported (e.g. for Fedex API details), but other shipping method settings should be part of the specific instance within a zone.

## My Wish List

- [ ] Deprecate current (legacy) Flat Rate Shipping
- [ ] Deprecate current (legacy) International Rate Shipping
- [ ] Deprecate current (legacy) Local Delivery
- [ ] Merge zones into core
- [ ] Redo the Zones UI as it's looking dated
- [ ] Have 2 default zones; "Everywhere Else" and "Everywhere"/"Worldwide". Rates inside "everywhere" apply to the world.
- [x] Introduce a new, simpler flat rate shipping method
- [ ] Make all shipping methods instance based so you can have multiple within each zone
- [ ] Work out the best way to store zones. Currently its custom tables (zones, zones_to_methods, zone_locations).
- [ ] Integrate new zones with the [onboarding wizard](https://github.com/woothemes/woocommerce/issues/7829)

## Backwards Compatibility

- Current shipping methods that are enabled need to be kept working
- TR shipping will need to detect the core zones UI and turn of its own
- Current shipping methods need to be placed in the 'everywhere'  zone.

## Over to you...

Really interested to get feedback from team and contributors. If possible, I'd like to make this the focus of 2.5/2.6.

@BFTrick Ideas board around this? What are other platforms doing?

@mattyza my proposal has business implications because we're essentially removing value from Table Rate shipping. On that side, I'd recommend merging Table Rate and Flat Rate Box shipping (and building volumetric shipping) into one "advanced shipping" plugin.

@roykho @claudiosmweb @jameskoster @barrykooij @kloon cc'ing for feedback

I'd also welcome fresh-eyed feedback from @davidlenehan and @justinshreve :)

**Comment by SiR-DanieL** [comment]
What do you mean here?

> Current shipping methods need to be placed in the 'everywhere' zone.

If i actually ship only to Italy, why would you place it in "Everywhere"?

What impact would it have on sites who are using Table Rate Shipping? They will need a sort of converter or something, unless the update of TRS comes together with WC.

**Comment by mikejolley** [comment]
@SiR-DanieL Shipping methods now, without zones, are 'everywhere'. That would be the BW compatible way to handle them, even if they have their own country-specific settings inside (which we will remove in non-legacy versions).

> What impact would it have on sites who are using Table Rate Shipping? They will need a sort of converter or something, unless the update of TRS comes together with WC.

Should be a straight swap unless we change data structure.

**Comment by mattyza** [comment]
+1 to the overall concept, the wishlist and the backwards compatibility, @mikejolley.

As for business implications, I'd say lets put together a game plan and then assess from there. As I see it, Table Rate Shipping still offers it's huge value proposition, only that the Zones part would be in WC core.

**Comment by SiR-DanieL** [comment]
> Shipping methods now, without zones, are 'everywhere'. That would be the BW compatible way to handle them, even if they have their own country-specific settings inside (which we will remove in non-legacy versions).

I think we should be careful with this.
It's true that a Flat rate is always Everywhere, even if it has contry specific settings inside, but those country specific settings should be converted into a zone without the user doing this manually.

**Comment by mikejolley** [comment]
@SiR-DanieL That would mean re-building the 'legacy' flat rate shipping method. Something I think we should avoid since its so damn bad right now. Not usable at all.

**Comment by SiR-DanieL** [comment]
And can we do something to help admins to move their restricted shipping methods to the new zones?
I mean not only good documentation.

I think they need to get used to the new interfaces and concept of zones. That could result in a temporary timeout of their businesses and a lot of tickets for the Shipping team.

**Comment by mikejolley** [comment]
@SiR-DanieL The plan above means existing shipping will not be affected; it will be moved to the everywhere zone. If a user takes no action, nothing changes on their store.

**Comment by SiR-DanieL** [comment]
:+1: i'm ok with that. The important is that this change does not break their business

**Comment by franticpsyx** [comment]
>If we were to merge zones into core Flat Rate Shipping would be so much more useful. Take out the bloat - just offer a single cost field and boom you could add multiple flat rates to multiple zones to make it much more flexible. "international shipping" could be dropped.

>my proposal has business implications because we're essentially removing value from Table Rate shipping. On that side, I'd recommend merging Table Rate and Flat Rate Box shipping (and building volumetric shipping) into one "advanced shipping" plugin

What Mike describes here was one of the things I wanted to take care of with Conditional Shipping and Payments. I've seen countless cases where Table Rate Shipping was overkill for what people wanted to achieve (conditional Flat Rates by country/weight). The extension, among tons of other things, makes Flat Rate Shipping more usable by allowing you to conditionally activate Flat Rates, for instance see http://docs.woothemes.com/document/woocommerce-conditional-shipping-and-payments/#section-5

I fully back up the solution that Mike recommends, in order to remove all the Flat Rate bloat and make things a bit more centralized. However, this is going to have direct implications in extensions such as CSP that need to be communicated in advance, since certain customer setups will no longer work after this finds its way into core.

**Comment by franticpsyx** [comment]
>Current shipping methods that are enabled need to be kept working
>TR shipping will need to detect the core zones UI and turn of its own
>Current shipping methods need to be placed in the 'everywhere' zone.

+1 from me on this one -- sounds like this will ensure that any extensions that rely on using the current Flat Rate ids will keep working (devs should only need to change the way Flat Rate ids are retrieved from instances). If the Flat Rate instances by default get added to an Everywhere Zone, then it will all be smooth sailing.

**Comment by roykho** [comment]
+1 on this and a solid plan.

**Comment by claudiosmweb** [comment]
+1

**Comment by BFTrick** [comment]
Hey Mike,

Thanks for bringing this up. This is definitely something that I think we should do at some point but I'm not sure if it's the right thing to do right now.

It's definitely powerful. I would love to give people in Denver free shipping, people in CO shipping for $10, and use USPS for the rest of the US. That is _massively_ powerful.

* What percentage of our users need this functionality?
* What percentage of our users are happy with Table Rate Shipping as it is? Or to ask this another way how many people would we help by putting it in core?
* Would a new user have any idea what a shipping zone is?
* What is our goal with this? To get more users who wouldn't get WC without this functionality or to help existing users who already use WC but could use with additional power?

# Extension Model

One of my favorite things about the extension model is that we can make the core product as lean and simple as humanly possible. There are plenty of people who use flat rate shipping as is and are perfectly happy with it. People who want the advanced functionality can search for it, learn about it, and use it.

It breaks the learning up into multiple steps. Introducing someone new to TRS is a bit of a learning process and it helps if they don't have to learn the rest of WooCommerce at the same time.

# Others

Shopify offers a [simplified version](https://docs.shopify.com/manual/settings/shipping/shipping-rates) of this. Basically select a country and then add a flat rate shipping option to it. What they do well is avoid jargon (something you have to learn and is hard for users).

What they don't do so well is let you combine different shipping methods. Ex. Use USPS for continental US and use $50 shipping to Alaska + Hawaii.

# Suggestion

What I think we should do is come up with a way to have shipping zones without calling them shipping zones and to make it visually intuitive. If we can do that it won't be much harder than setting up flat rate shipping.

If we simply merge shipping zones into WC core as is we'll confuse our users.

As always we have to compare this idea against other ideas for 2.5. I don't think GitHub is the place to discuss which features are better but there are definitely some other powerful features to consider ex. backend ordering, more stock reports, etc.

**Comment by mikejolley** [comment]
@BFTrick this idea has been floating around for 2 years FYI ;) Shipping in it's current form is probably the most confusing part of WC. I think this deserves a look at.

I don't mind if we don't call it zones, and we certainly don't need to keep the TRS UI. The structure and model however can come across, and 'zone' can be the terminology in the code.

**Comment by franticpsyx** [comment]
>The structure and model however can come across, and 'zone' can be the terminology in the code.

I think the main idea is to allow people to select countries for each additional Flat Add-On rate they create. Right?

>What I think we should do is come up with a way to have shipping zones without calling them shipping zones and to make it visually intuitive. If we can do that it won't be much harder than setting up flat rate shipping.

I agree with Patrick that there's no need to use the TRS nomenclature of Zones for something so simple. Same with the UI. TRS is just too complex for the majority of cases where it's actually used.

I'd give a :+1: to a simple UI that will allow people to create as many Flat-Rate instances as they need, and allow them to select a different group of Countries / States for each one. 

That would probably have serious business implications for TRS and CSP, but at the moment, both TRS and (to a lesser extent) CSP give the impression of having to get a bazooka in order to kill an ant.

**Comment by corsonr** [comment]
I think that there's a important part of WC users complaining about the few features provided by the actual flat rate shipping method. Here is a sample here #8380.

So i'm a total fan of this proposition and as WC support shipping team member I can provide details and customers expectations if needed.

**Comment by theleemon** [comment]
+1

**Comment by xrelaxerx** [comment]
+1 on adding Shipping Zones support to core.

**Comment by mikejolley** [comment]
Closing in favour of #9826
